csharp/src/Microsoft.ML.OnnxRuntime/NativeOnnxTensorMemory.cs

```diff
@@ -172,54 +162,26 @@ public override MemoryHandle Pin(int elementIndex = 0)
                 {
                     throw new ArgumentOutOfRangeException(nameof(elementIndex));
                 }
-                Retain();
-
                 return new MemoryHandle((void*)((int)_dataBufferPointer + elementIndex * _elementWidth)); //could not use Unsafe.Add
             }
         }
 
-        public override void Unpin()
-        {
-            Release();
-        }
-
-        private bool Release()
-        {
-            int newRefCount = Interlocked.Decrement(ref _referenceCount);
-
-            if (newRefCount < 0)
-            {
-                throw new InvalidOperationException("Unmatched Release/Retain");
-            }
-
-            return newRefCount != 0;
-        }
+        // MemoryHandle returned above should unpinned when done.
+        public override void Unpin() { }
 
```

Comment:
The comments for Pin and Unpin seem a bit confusing. The underlying memory is unmanaged so the address is never going to change is it? As such there's no actual 'pin' involved - it's just that the value returned was already pinned, so Unpin is a no-op (and MemoryHandle should be disposed when done - it doesn't have an Unpin method).

Reply:
This class needs work. My understanding that this class takes managed Tensor memory, pins it and then provides as input to OrtValue.

---
In reply to: [476843544](https://github.com/microsoft/onnxruntime/pull/4915#discussion_r476843544) [](ancestors = 476843544)